### PR TITLE
Remove deprecated Threads functionalities

### PR DESCRIPTION
**Pull request by masterleinad**
Deprecated in #13116 and #12173.

### CODE REVIEW

File: tests/base/generate_normal_random_number_01.cc [hunk 1]
```diff
@@ -54,11 +54,11 @@ main()
   initlog();
 
   // create 4 threads, run the test function on them
-  Threads::Thread<std::pair<double, double>> tg[4];
-  tg[0] = Threads::new_thread(&test);
-  tg[1] = Threads::new_thread(&test);
-  tg[2] = Threads::new_thread(&test);
-  tg[3] = Threads::new_thread(&test);
+  std::async tg[4] = {};
```
Comment (bangerth) [inline tests/base/generate_normal_random_number_01.cc:57]: Wait, `std::async` is a function. How does this work? Did you mean this to be
```suggestion
  std::vector<std::future> tg;
```
?

File: tests/multithreading/thread_local_storage_05.cc [hunk 2]
```diff
@@ -34,7 +34,7 @@ Threads::ThreadLocalStorage<X> tls_data;
 
 
 void
-execute(Threads::Mutex &m)
+execute(Threads::Mutex *m)
```
Comment (bangerth) [inline tests/multithreading/thread_local_storage_05.cc:37]: Why is this change necessary?

File: tests/multithreading/thread_local_storage_06.cc [hunk 3]
```diff
@@ -46,7 +46,7 @@ Threads::ThreadLocalStorage<X> tls_data(forty_two);
 
 
 void
-execute(Threads::Mutex &m)
+execute(Threads::Mutex *m)
```
Comment (bangerth) [inline tests/multithreading/thread_local_storage_06.cc:49]: same here, and perhaps other tests?

**Review by bangerth — CHANGES_REQUESTED** [review]

File: tests/base/generate_normal_random_number_01.cc (hunk 1 above)
Comment (masterleinad) [reply tests/base/generate_normal_random_number_01.cc]: Fixed now.

File: tests/multithreading/thread_local_storage_05.cc (hunk 2 above)
Comment (masterleinad) [reply tests/multithreading/thread_local_storage_05.cc]: Fixed differently, but the issue is that `std::thread` copies its arguments even if they are meant to be references.

File: tests/multithreading/thread_local_storage_06.cc (hunk 3 above)
Comment (masterleinad) [reply tests/multithreading/thread_local_storage_06.cc]: I'm using `std::ref` here now and also in other places.

**Review by bangerth — APPROVED** [review]
